File: leetsync/2456-construct-smallest-number-from-di-string/construct-smallest-number-from-di-string.java
class Solution {
    public String smallestNumber(String pattern) {
        int n = pattern.length();
        String num = "";
        Stack<Character> stack = new Stack<>();
        int counter = 1;
        for (int i = 0; i <= n; i++) {
            stack.push((char) ('0' + counter));
            counter++;
            if (i == n || pattern.charAt(i) == 'I') {
                while (!stack.isEmpty()) {
                    num += stack.pop();
                }
            }
        }
        return num;
    }
}

File: leetsync/2456-construct-smallest-number-from-di-string/construct-smallest-number-from-di-string.go
func smallestNumber(pattern string) string {
	n := len(pattern)
	num := make([]byte, n+1)
	stack := make([]byte, 0)
	counter := byte('1')
	for i := 0; i <= n; i++ {
		stack = append(stack, counter)
		counter++
		if i == n || pattern[i] == 'I' {
			for len(stack) > 0 {
				top := len(stack) - 1
				num[i-top] = stack[top]
				stack = stack[:top]
			}
		}
	}
	return string(num)
}